indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java

```diff
@@ -773,4 +790,123 @@ public void setInvalidRowCount(long invalidRowCount)
+
+    private int counter;
+    private int checkInterval = 20000;
+    private int maxRowsInMemory = 500000;
+
+    @Override
+    protected void setup(Context context) throws IOException, InterruptedException
+    {
+      super.setup(context);
+      aggregators = config.getSchema().getDataSchema().getAggregators();
+      combiningAggs = new AggregatorFactory[aggregators.length];
+      for (int i = 0; i < aggregators.length; ++i) {
+        combiningAggs[i] = aggregators[i].getCombiningFactory();
+      }
+    }
+
+    @Override
+    protected void cleanup(Context context) throws IOException, InterruptedException
+    {
+      for (IncrementalIndex index : combiners.values()) {
+        flush(context, index);
+      }
+      combiners.clear();
+    }
+
+    @Override
+    protected void innerMap(InputRow inputRow, Object value, Context context)
+        throws IOException, InterruptedException
+    {
+      final long timestamp = granularitySpec.getQueryGranularity().truncate(inputRow.getTimestampFromEpoch());
+
+      IncrementalIndex target = combiners.get(timestamp);
+      if (target == null) {
+        combiners.put(timestamp, target = makeIncrementalIndex(timestamp, aggregators, config, null));
+      }
+
+      if (!target.canAppendRow()) {
+        flush(context, target);
+        combiners.put(timestamp, target = makeIncrementalIndex(timestamp, aggregators, config, null));
+      }
+      target.add(inputRow);
+
+      if (++counter % checkInterval == 0) {
+        flushOldestIfNeeded(context);
+      }
+    }
+
+    private void flushOldestIfNeeded(Context context) throws IOException, InterruptedException
+    {
+      while (totalRowsInMemory() > maxRowsInMemory) {
+        IncrementalIndex oldest = null;
+        for (IncrementalIndex index : combiners.values()) {
+          if (oldest == null || index.getLastAccessTime() < oldest.getLastAccessTime()) {
+            oldest = index;
+          }
+        }
+        if (oldest != null) {
+          flush(context, oldest);
+          combiners.remove(oldest.getInterval().getStartMillis());
+        }
+      }
+    }
+
+    private void flush(Context context, IncrementalIndex index) throws IOException, InterruptedException
+    {
+      for (Object row : index) {
+
+        InputRow inputRow = (InputRow) row;
+        // Group by bucket, sort by timestamp
+        final Optional<Bucket> bucket = getConfig().getBucket(inputRow);
+
+        if (!bucket.isPresent()) {
+          throw new ISE("WTF?! No bucket found for row: %s", inputRow);
+        }
+
+        final byte[] hashedDimensions = hashFunction.hashBytes(
+            HadoopDruidIndexerConfig.JSON_MAPPER.writeValueAsBytes(
+                Rows.toGroupKey(
+                    inputRow.getTimestampFromEpoch(),
+                    inputRow
+                )
+            )
+        ).asBytes();
+
+        byte[] serializedInputRow = InputRowSerde.toBytes(inputRow, combiningAggs);
```

Comment:
Not related with this patch, I'm curious about the way of `toBytes()` working.

In this case, inputRow already has aggregated value for each aggregator in itself.
However, `toBytes()` makes combining aggregator and compute aggregation with that aggregated value, 
which results in the same value as input.
I'm not sure making aggregator and doing meaningless aggregation induces how much overhead but I think it is unnecessary.